plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+
+        std::vector<std::vector<NodeID>> route_result;
+        route_result.reserve(scc.GetNumberOfComponents());
+        TIMER_START(trip);
+        //run Trip computation for every SCC
+        for (std::size_t k = 0; k < scc.GetNumberOfComponents(); ++k) {
+            const auto component_size = scc.range[k+1] - scc.range[k];
+
+            BOOST_ASSERT_MSG(component_size >= 0,"invalid component size");
+
+            if (component_size > 1) {
+                std::vector<NodeID> scc_route;
+                NodeIDIterator start = std::begin(scc.component) + scc.range[k];
+                NodeIDIterator end = std::begin(scc.component) + scc.range[k+1];
+
+                if (component_size < BF_MAX_FEASABLE) {
+                    scc_route = osrm::trip::BruteForceTrip(start, end, number_of_locations, result_table);
+                    route_result.push_back(scc_route);
+                } else {
+                    scc_route = osrm::trip::FarthestInsertionTrip(start, end, number_of_locations, result_table);
+                    route_result.push_back(scc_route);
+                }
+
+                // use this if output if debugging of route is needed:
+                SimpleLogger().Write() << "Route #"
+                                       << k << ": "
+                                       << [&scc_route](){
+                                               std::string s = "";
+                                               for (auto x : scc_route) {
+                                                   s += std::to_string(x) + " ";
+                                               }
+                                               return s;
+                                           }();
+            } else {
+                // if component only consists of one node, add it to the result routes
+                route_result.push_back({scc.component[scc.range[k]]});
+            }
+        }
+
+        // compute all round trip routes
+        std::vector<InternalRouteResult> comp_route (route_result.size());
+        for (std::size_t r = 0; r < route_result.size(); ++r) {
+            ComputeRoute(phantom_node_vector, route_parameters, route_result[r], comp_route[r]);
+        }
+
+        TIMER_STOP(trip);
+
+        // prepare JSON output
+
+
+        // create a json object for every trip
+        osrm::json::Array trip;
+        for (std::size_t i = 0; i < route_result.size(); ++i) {
+            std::unique_ptr<BaseDescriptor<DataFacadeT>> descriptor;
+            descriptor = osrm::make_unique<JSONDescriptor<DataFacadeT>>(facade);
+            descriptor->SetConfig(route_parameters);
+
+            osrm::json::Object scc_trip;
+
+            // set permutation output
+            SetLocPermutationOutput(route_result[i], scc_trip);
+            // set viaroute output
+            descriptor->Run(comp_route[i], scc_trip);
+
+            trip.values.push_back(scc_trip);
+        }
+
+        json_result.values["trips"] = trip;
```

Comment:
`trip` is not needed anymore --- if this is a map, you can `.insert` and then `std::move()` it in, as a map's `operator[]` does not provide an rvalue overload.